examples/static-gridLSTM/LSTM.hs

```diff
@@ -0,0 +1,112 @@
+                                                , SomeException(..)
+                                                )
+import           Data.List                      ( foldl'
+                                                , intersperse
+                                                , scanl'
+                                                )
+import           Data.Reflection
+import           Foreign.ForeignPtr
+import           GHC.Generics
+import           GHC.TypeLits
+import           GHC.TypeLits.Extra
+import           System.Environment
+import           System.IO.Unsafe
+import           Prelude                 hiding ( tanh )
+import qualified ATen.Cast                     as ATen
+import qualified ATen.Class                    as ATen
+import qualified ATen.Type                     as ATen
+import qualified ATen.Managed.Type.Tensor      as ATen
+import qualified Torch.Autograd                as A
+import qualified Torch.DType                   as D
+import qualified Torch.Functions               as D
+import qualified Torch.NN                      as A
+import qualified Torch.Tensor                  as D
+import qualified Torch.TensorFactories         as D
+import           Torch.Typed
+import           Torch.Typed.Factories
+import           Torch.Typed.Native      hiding ( linear )
+import           Torch.Typed.NN
+import           Torch.Typed.NN.Recurrent.LSTM
+
+run
+    :: (KnownDType d, KnownNat i, KnownNat h, KnownNat b)
+    => [(Tensor d '[b, i])]
+    -> Tensor d '[b, h]
+    -> LSTMCell d i h
+    -> Int
+    -> IO (LSTMCell d i h)
+run input_tensors expected_output model i = do
+
+    let Just output = snd <$> forward model input_tensors
+    let loss        = mse_loss output expected_output
+
+    print loss
+
+    let flat_parameters = A.flattenParameters model
+    let gradients       = A.grad (toDynamic loss) flat_parameters
+    -- new parameters returned by the SGD update functions
+    new_flat_parameters <- mapM A.makeIndependent
+        $ A.sgd 5e-2 flat_parameters gradients
+
+    -- return the new model state "to" the next iteration of foldLoop
+    return $ A.replaceParameters model new_flat_parameters
+
+toBackend
+    :: forall t . (ATen.Castable t (ForeignPtr ATen.Tensor)) => String -> t -> t
+toBackend backend t = unsafePerformIO $ case backend of
+    "CUDA" -> ATen.cast1 ATen.tensor_cuda t
+    _      -> ATen.cast1 ATen.tensor_cpu t
+
+main :: IO ()
+main = do
+    backend' <- try (getEnv "BACKEND") :: IO (Either SomeException String)
+    let backend = case backend' of
+            Right "CUDA" -> "CUDA"
+            _            -> "CPU"
+    let foldLoop x count block = foldM @[] block x [1 .. count]
+
+    -- randomly initializing training values
+    (input_tensor :: [Tensor D.Float '[1, 2]]) <- mapM
+        (const (randn :: IO (Tensor D.Float '[1, 2])))
+        [0 .. 4]
+    expected_output                 <- randn :: IO (Tensor D.Float '[1, 2])
```

Comment:
hmm, any idea what the expected behaviour of an lstm cell is when you train it on normally distributed input and output data?

Reply:
This is just a copy of the memorization task that exists for the current recurrent models in example.  It was only used to make sure `autograd` was sensible.